controllers/metal3.io/baremetalhost_controller.go

```diff
@@ -1112,41 +1121,28 @@ func saveHostProvisioningSettings(host *metal3v1alpha1.BareMetalHost) (dirty boo
 	}
 
 	// Copy RAID settings
-	if host.Spec.RAID != host.Status.Provisioning.RAID {
-		// If RAID settings is nil, remove saved settings,
-		// else check hardware RAID and software RAID.
-		if host.Spec.RAID == nil {
-			host.Status.Provisioning.RAID = nil
-			dirty = true
-		} else {
-			if host.Status.Provisioning.RAID == nil {
-				host.Status.Provisioning.RAID = &metal3v1alpha1.RAIDConfig{}
-				dirty = true
-			}
-			// If HardwareRAIDVolumes isn't nil, we will ignore SoftwareRAIDVolumes.
-			if len(host.Spec.RAID.HardwareRAIDVolumes) != 0 {
-				// If software RAID has been saved, remove it.
-				if len(host.Status.Provisioning.RAID.SoftwareRAIDVolumes) != 0 {
-					host.Status.Provisioning.RAID.SoftwareRAIDVolumes = nil
-				}
-				// Compare hardware RAID settings
-				if !reflect.DeepEqual(host.Spec.RAID.HardwareRAIDVolumes, host.Status.Provisioning.RAID.HardwareRAIDVolumes) {
-					host.Status.Provisioning.RAID.HardwareRAIDVolumes = host.Spec.RAID.HardwareRAIDVolumes
-					dirty = true
-				}
-			} else {
-				// If hardware RAID has been saved, remove it.
-				if len(host.Status.Provisioning.RAID.HardwareRAIDVolumes) != 0 {
-					host.Status.Provisioning.RAID.HardwareRAIDVolumes = nil
-					dirty = true
-				}
-				// Compare software RAID settings
-				if !reflect.DeepEqual(host.Spec.RAID.SoftwareRAIDVolumes, host.Status.Provisioning.RAID.SoftwareRAIDVolumes) {
-					host.Status.Provisioning.RAID.SoftwareRAIDVolumes = host.Spec.RAID.SoftwareRAIDVolumes
-					dirty = true
-				}
-			}
+	specRAID := host.Spec.RAID.DeepCopy()
+	// If RAID configure is nil or empty, means that we need to keep the current hardware RAID configuration
+	// or clear current software RAID configuration
+	if specRAID == nil || reflect.DeepEqual(specRAID, &metal3v1alpha1.RAIDConfig{}) {
+		// Set the default value of RAID configure:
+		// {
+		//     HardwareRAIDVolumes: nil or Status.Provisioning.RAID.HardwareRAIDVolumes(not empty),
+		//     SoftwareRAIDVolume: [],
+		// }
+		specRAID = &metal3v1alpha1.RAIDConfig{}
+		if host.Status.Provisioning.RAID != nil && len(host.Status.Provisioning.RAID.HardwareRAIDVolumes) != 0 {
+			specRAID.HardwareRAIDVolumes = host.Status.Provisioning.RAID.DeepCopy().HardwareRAIDVolumes
 		}
+		specRAID.SoftwareRAIDVolumes = []metal3v1alpha1.SoftwareRAIDVolume{}
+	}
+	// If HardwareRAIDVolumes isn't nil, ignore SoftwareRAIDVolumes
+	if specRAID.HardwareRAIDVolumes != nil {
+		specRAID.SoftwareRAIDVolumes = nil
```

Comment:
Our logic time ignores `softwaraRAIDVolumes` when `hardwareRAIDVolumes` isn't nil, so I think softwareRAIDVolume can not affect the value of dirty flag in this case.

Reply:
Our logic time ignores `softwaraRAIDVolumes` when `hardwareRAIDVolumes` isn't nil, so I think softwareRAIDVolume can not affect the value of dirty flag in this case.

Reply:
> In addition I'd expect this method to modify just the Status

I just modified the copy of the spec, which will not affect the actual content of the spec.

Reply:
Because `BuildRAIDCleanSteps` will also handle the situation where software RAID and hardware RAID exist at the same time, so I removed the logic to handle this situation from `saveHostProvisioningSettings` to avoid `DeepCopy`.